### PR TITLE
Trending Carousel missing default bookcovers

### LINKED ISSUE #6648: Trending Carousel missing default bookcovers
This is a fairly prominently displayed feature on the homepage so it would be good to get this out quickly if the fix is easy

### Evidence / Screenshot (if possible)

openlibrary org_ (2)

### Steps to Reproduce

1. Go to ...
2. Do ...

* Actual:
* Expected:

### Details

- **Logged in (Y/N)?**
- **Browser type/version?**
- **Operating system?**
- **Environment (prod/dev/local)?** prod

### PR DESCRIPTION
**Pull request by vchapandrews**
Closes #6648

 fix

### Technical

I believe the problem exists:

- On [line 15](https://github.com/internetarchive/openlibrary/pull/6722#discussion_r917009198) (bug: the conditional statement should not be satisfied when `book.get('cover_id')` returns the string value, 'undefined')
- ...which is resulting in another bug on [line 18](https://github.com/internetarchive/openlibrary/blob/71a074f49b64e391ed92022668488a6fdec11d5d/openlibrary/templates/books/custom_carousel.html#L18) in **openlibrary/templates/books/custom_carousel.html** (as we are passing an undefined value into our course_url string - which leads to a "dead" link i.e. no image)

### Explanation

From looking in browser developer console (shown below), it seems [`book.get('cover_id')`](https://github.com/internetarchive/openlibrary/blob/71a074f49b64e391ed92022668488a6fdec11d5d/openlibrary/templates/books/custom_carousel.html#L18) and/or [`book.get('cover_i')`](https://github.com/internetarchive/openlibrary/blob/71a074f49b64e391ed92022668488a6fdec11d5d/openlibrary/templates/books/custom_carousel.html#L18) are returning the string value, 'undefined', which is ending up in the image `src=" "` attribute [here](https://github.com/internetarchive/openlibrary/blob/71a074f49b64e391ed92022668488a6fdec11d5d/openlibrary/templates/books/custom_carousel.html#L45).

See below:

If you _right-click_ then choose _Inspect_ on the items with missing default bookcovers, they all seem to have `src="//covers.openlibrary.org/b/id/undefined-M.jpg"` (an invalid link).

image

image

image

# Proposed solution:

There may be a better way to solve this, but a quick temporary fix is to add a conditional statement that sets `cover_url` to **`False`** if `book.get('cover_id')` returns a string value of 'undefined' or anything that isn't a _valid_ cover id.

By setting `cover_url` to **`False`**, it means that [this conditional statement on line 37](https://github.com/internetarchive/openlibrary/blob/71a074f49b64e391ed92022668488a6fdec11d5d/openlibrary/templates/books/custom_carousel.html#L37) performs correctly (i.e. only items with valid cover_urls should satisfy the condition, and all other items are dealt with by the else statement which creates a default cover.)

Before:
image

After:
image

### Testing

Note, this hasn't been tested as I'm not sure how to get the trending carousel to show up in the development mode (plus I'm only seeing like 6 books there .)

Is there a way to run the site as it is online (i.e. not development version), but in my local environment? If so, I'll test this solution.

### CODE REVIEW

File: openlibrary/templates/books/custom_carousel.html [hunk 1]
```diff
@@ -15,6 +15,8 @@
   $if book.get('cover_url'):
     $ cover_url = book.get('cover_url')
   $elif book.get('cover_id') or book.get('cover_i'):
```
Comment (vchapandrews) [inline openlibrary/templates/books/custom_carousel.html:17]: This condition is designed to only be satisfied if a value for 'cover_id' is found i.e. if `book.get('cover_id')` returns a value that evaluates to **TRUE**.

However, for some books that don't have a cover ID, `book.get('cover_id')` is returning the string **'undefined'**. And as 'undefined' technically evaluates to **TRUE**, the code inside this condition is running, even though no valid cover id has been provided.

I believe that's the root cause of this issue: #6648

My proposed solution is here: #6722

**Comment by mekarpeles** [comment]
Much of the template code is checking variables in pyhton (not js) so I presume the value `undefined` will not come up often `openlibrary/templates/books/custom_carousel.html`.

**Review by mekarpeles — CHANGES_REQUESTED** [review]
Much of the template code is checking variables in python (not js) so I presume the value `undefined` will not come up often `openlibrary/templates/books/custom_carousel.html`.

When a book carousel is first loaded on openlibrary.org, it uses custom_carousel.html to fetch and render some ~20 or so works. This carousel (during its instantiation) specifies a `loadMore` url which is capable of fetching additional json data from API endpoints and then rendering each of the book items returned by the API as an item which may be added to the carousel.

This logic happens here:
https://github.com/internetarchive/openlibrary/blob/db2d35ad264adf6edcffbfb72f2ecdea304fba50/openlibrary/plugins/openlibrary/js/carousel/Carousel.js#L84-L177. Because this is done in javascript, presumably the cover is appearing as `undefined` in the javascript.

**Comment by jimchamp** [comment]
I agree with @mekarpeles.  It's probably best to construct the image source URL and add that to the `` inside of `$el` via interpolation.

You can use this as the default image source:  `/images/icons/avatar_book-lg.png`

Regarding local testing:
I've noticed that this issue happens all carousels that display book covers, so no need to get the trending carousel working locally if it's not already.  If you're not seeing any book carousels at all, you can add one to the homepage for testing purposes, here: https://github.com/internetarchive/openlibrary/blob/master/openlibrary/templates/home/index.html

If you add `$:macros.QueryCarousel(query="a")` to that file, it will display a carousel containing the results of a search for the string "a" (not a useful search in production, but should yield a couple dozen hits locally).  Our homepage is cached, so you'll have to restart web and memcached after adding the carousel (just run `docker-compose restart web memcached` to do so).

**Comment by vchapandrews** [comment]
Thanks for pointing me in the right direction @mekarpeles

Is this what you had in mind @jimchamp?

```diff

+ const default_cover_url = "/images/icons/avatar_book-lg.png"
  
  const $el = $(`
      <div class="book carousel__item">
          <div class="book-cover">
              <a href="${work.key}">
-               <img class="bookcover" src="//covers.openlibrary.org/b/${cover.type}/${cover.id}-M.jpg">
+               <img class="bookcover" src="//covers.openlibrary.org/b/${cover.type}/${cover.id}-M.jpg?default=https://openlibrary.org${default_cover_url}">
              </a>
          </div>
          <div class="book-cta">
              <a class="btn cta-btn ${cls}"
                 data-ol-link-track="subjects"
                 data-key="subjects"
             >${cta}</a>
          </div>
      </div>`);
  $el.find('.bookcover').attr('title', work.title);
  $el.find('.cta-btn')
      .attr('title', `${cta}: ${work.title}`)
      .attr('data-ocaid', ocaid)
      .attr('href', url);
  return $el;
```


Adding the `?default=https://openlibrary.org/images/icons/avatar_book-lg.png` to img src seems to be working when I do it manually in developer console on the live site. See screenshots:

image
image

But when I make this change in the carousel.js file, nothing changes in my local environment. So seems there's something I'm missing here

**Comment by jimchamp** [comment]
> Adding the ?default=https://openlibrary.org/images/icons/avatar_book-lg.png to img src seems to be working [...]
TIL that we can send a default to our covers service.

Your first proposal, where `src=${cover_url}` looks perfect!

Any changes to `.js` files require the Javascript to be compiled again.  You can do this by running the following in your terminal:
`docker-compose exec web make js`
Some more details about this can be found [here](https://github.com/internetarchive/openlibrary/wiki/Frontend-Guide#css-js-and-html).

**Comment by vchapandrews** [comment]
Hey @jimchamp,

I've made the changes outlined above in this commit: 66cb8f6a3e1f5cb6bd2927f49cd347d44404d48 — which results in `/images/icons/avatar_book-lg.png` being displayed for books where cover.id returns a **falsy** value such as `undefined`. See below:

image

## Improvements

I have since updated this so that instead of just displaying the empty cover image (as shown above), it now also displays the book title and author (using the same method and html classes in custom_carousel.html.) See below:

image

Fundamentally works the same way as the previous version (66cb8f6a3e1f5cb6bd2927f49cd347d44404d48), but instead of just building the img `src` url, we're building out a block of html (similar to how [this block of code](https://github.com/internetarchive/openlibrary/blob/238ed8701d70a956e1074cb80e864da5ed01825b/openlibrary/templates/books/custom_carousel.html#L37-L51) works in custom_carousel.html)

I think this is worthwhile as the book titles aren't displayed anywhere else in the carousel (unless you hover over a book, which on mobile you can't do anyway.) So makes navigation at lot easier.

However, if you'd rather have it as it was, let me know and I'll roll back the pull request to previous commit.

(Also, thanks for the docker tips, everything is working now!)

File: openlibrary/templates/books/custom_carousel.html [hunk 2]
```diff
@@ -44,7 +44,7 @@
           $else:
             src="$(cover_url)"/>
         $else:
-          <div class="carousel__item__blankcover">
+          <div class="carousel__item__blankcover"  title="$(title)">
```
Comment (jimchamp) [inline openlibrary/templates/books/custom_carousel.html:47]: This fixes `title` not being present for initially loaded books that are missing cover images.  Good catch!

File: openlibrary/plugins/openlibrary/js/carousel/Carousel.js [hunk 3]
```diff
@@ -108,13 +108,30 @@ const Carousel = {
...
+                        <div class="carousel__item__blankcover bookcover">
+                            <div class="carousel__item__blankcover--title">${work.title}</div>
+                            ${work.author_name ? `<div class="carousel__item__blankcover--authors">${work.author_name}</div>` : ''}
+                        </div>
+                    <a/>`
+            }
+
+            const $el = $(`
+                <div class="book carousel__item">
+                    ${book_cover}
```
Comment (jimchamp) [inline openlibrary/plugins/openlibrary/js/carousel/Carousel.js:134]:
```suggestion
                <div class="book carousel__item">
                    <div class="book-cover">
                        ${book_cover}
                    </div>
```
The `.book-cover` div is present when the cover image is missing.

File: openlibrary/plugins/openlibrary/js/carousel/Carousel.js [hunk 4]
```diff
@@ -108,13 +108,30 @@ const Carousel = {
...
 
-            const $el = $(`
-                <div class="book carousel__item">
+
+            const default_cover_url = 'https://openlibrary.org/images/icons/avatar_book-lg.png';
+
+            let book_cover = undefined
+            if (cover.id) {
+                book_cover = `
                     <div class="book-cover">
```
Comment (jimchamp) [inline openlibrary/plugins/openlibrary/js/carousel/Carousel.js:117]: Remove this div from `bookCover`.  It should be present for all cases, so add it to `$el`.
Don't forget to remove the closing div here, as well.

File: openlibrary/plugins/openlibrary/js/carousel/Carousel.js [hunk 5]
```diff
@@ -108,13 +108,30 @@ const Carousel = {
...
                         <a href="${work.key}">
-                            <img class="bookcover" src="//covers.openlibrary.org/b/${cover.type}/${cover.id}-M.jpg">
+                            <img class="bookcover" src="//covers.openlibrary.org/b/${cover.type}/${cover.id}-M.jpg?default=${default_cover_url}">
```
Comment (jimchamp) [inline openlibrary/plugins/openlibrary/js/carousel/Carousel.js:119]:
```suggestion
                            <img class="bookcover" src="//covers.openlibrary.org/b/${cover.type}/${cover.id}-M.jpg">
```

**Review by jimchamp — CHANGES_REQUESTED** [review]
This is a great approach, @vchapandrews!
Now that the is being rendered the same the same way client-side and in our template, I'm not seeing the need to explicitly set a default cover image.

I've deployed this code to https://testing.openlibrary.org , and it's looking good.  I'm happy to merge this PR once the requested changes have been made.  Nice work!

File: openlibrary/plugins/openlibrary/js/carousel/Carousel.js [hunk 6]
```diff
@@ -108,13 +108,30 @@ const Carousel = {
...
                         </a>
-                    </div>
+                    </div>`
+            } else {
+                book_cover = `
```
[the remark this answers is not in the document]
Comment (vchapandrews) [reply openlibrary/plugins/openlibrary/js/carousel/Carousel.js]: Have updated all instances of 'book_cover' to 'bookCover'

File: openlibrary/plugins/openlibrary/js/carousel/Carousel.js [hunk 7]
```diff
@@ -108,13 +108,30 @@ const Carousel = {
                 cover.id = ocaid;
             }
 
-            const $el = $(`
-                <div class="book carousel__item">
+
+            const default_cover_url = 'https://openlibrary.org/images/icons/avatar_book-lg.png';
+
+            let book_cover = undefined
```
[the remark this answers is not in the document]
Comment (vchapandrews) [reply openlibrary/plugins/openlibrary/js/carousel/Carousel.js]: Have updated all instances of 'book_cover' to 'bookCover'

File: openlibrary/plugins/openlibrary/js/carousel/Carousel.js (hunk 4 above)
Comment (vchapandrews) [reply openlibrary/plugins/openlibrary/js/carousel/Carousel.js]: Have moved both `<div class="book-cover">` and `<a href="${work.key}">` to the $el tag. (Moved the `<a href="${work.key}"` tag as was identical in both the image book_cover path and the default cover path.)

**Comment by vchapandrews** [comment]
> This is a great approach, @vchapandrews! Now that the is being rendered the same the same way client-side and in our template, I'm not seeing the need to explicitly set a default cover image.
> 
> I've deployed this code to https://testing.openlibrary.org/ , and it's looking good. I'm happy to merge this PR once the requested changes have been made. Nice work!

Hey @jimchamp,

Glad to hear it's all working in the test deployment and that you're happy with the changes!

Regarding the default cover image, I actually noticed a strange issue which is why I've left `?default=` in for now.

Curious to hear what your thoughts are on this, as I've only really noticed this issue in my local environment (not on the actual live website), but I guess in theory it _could_ happen on the live website too?

### The problem

Essentially, I've noticed that in some cases a book can have a cover image that is damaged/corrupted (or for whatever reason does not display properly.) This results in a missing bookcover for that book, because the image linked in `img src=` does not open properly.

Ideally, these "broken" images should **fail** the [`if (cover.id)`](https://github.com/internetarchive/openlibrary/pull/6722/files?w=0#diff-c3934c40bc5db82d9d98398a7d6d6e7b6642ef17ddad13b78d3799f09a755d1fR112) condition (on [line 112](https://github.com/internetarchive/openlibrary/pull/6722/files?w=0#diff-c3934c40bc5db82d9d98398a7d6d6e7b6642ef17ddad13b78d3799f09a755d1fR112)) so that a default cover is generated.

But because there technically is a cover image (albeit a broken one), the [`if (cover.id)`](https://github.com/internetarchive/openlibrary/pull/6722/files?w=0#diff-c3934c40bc5db82d9d98398a7d6d6e7b6642ef17ddad13b78d3799f09a755d1fR112) condition (on [line 112](https://github.com/internetarchive/openlibrary/pull/6722/files?w=0#diff-c3934c40bc5db82d9d98398a7d6d6e7b6642ef17ddad13b78d3799f09a755d1fR112)) evaluates to `True` in this scenario, and the site ends up displaying an invalid image.

### Possible solutions

The `?default=https://openlibrary.org/images/icons/avatar_book.png` seems to be able to catch these "broken" images and replace them with the default cover we specify.

That's the only reason why it might be worth keeping it. But again, I haven't really noticed this issue on the website. So might be a case of fixing something that's not broken.

Ideally, we'd be able to catch these broken images directly in our script and ensure that a default cover with title and author is generated in these cases (instead of just a placeholder image.)

---

Anyway, if you want me to remove it let me know, but just thought I should bring this to your attention

**Review by jimchamp — APPROVED** [review]
I've deployed this to testing, and everything is working as expected.  Thanks @vchapandrews!

**Pull request merged**